### ISSUE TITLE
ICE on develop with gcc8

### ISSUE THREAD

**Issue by fmahebert** [sxs-collaboration/spectre#1272]
# Bug reports:

### Expected behavior:

Code compiles.

### Current behavior:

On my desktop with gcc8, after #970 was merged, I get the following ICE:

```
In file included from /home/francois/code/spectre-work/spectre/tests/Unit/Elliptic/Initialization/Test_Derivatives.cpp:16:
/home/francois/code/spectre-work/spectre/src/DataStructures/Variables.hpp: In substitution of ‘template<class ... WrappedTags, typename Requires_detail::requires_impl<flat_all_v<is_same_v<typename db::DataBox_detail::remove_all_prefixes_impl<WrappedTags, is_base_of_v<db::PrefixTag, WrappedTags> >::type, typename db::DataBox_detail::remove_all_prefixes_impl<Tags, is_base_of_v<db::PrefixTag, Tags> >::type>...> >::template_error_type_failed_to_meet_requirements_on_template_parameters <anonymous> > Variables<brigand::list<Tags::deriv<{anonymous}::ScalarFieldTag, brigand::integral_constant<long unsigned int, 2>, Frame::Inertial, std::nullptr_t> > >& Variables<brigand::list<Tags::deriv<{anonymous}::ScalarFieldTag, brigand::integral_constant<long unsigned int, 2>, Frame::Inertial, std::nullptr_t> > >::operator=<WrappedTags ..., <enumerator> >(const Variables<brigand::list<Ts ...> >&) [with WrappedTags = {Tags::deriv<{anonymous}::ScalarFieldTag, brigand::integral_constant<long unsigned int, 2>, Frame::Inertial, std::nullptr_t>}; typename Requires_detail::requires_impl<flat_all_v<is_same_v<typename db::DataBox_detail::remove_all_prefixes_impl<WrappedTags, is_base_of_v<db::PrefixTag, WrappedTags> >::type, typename db::DataBox_detail::remove_all_prefixes_impl<Tags, is_base_of_v<db::PrefixTag, Tags> >::type>...> >::template_error_type_failed_to_meet_requirements_on_template_parameters <anonymous> = <missing>]’:
/home/francois/code/spectre-work/spectre/src/DataStructures/DataBox/Deferred.hpp:168:8:   required from ‘void Deferred_detail::deferred_assoc_state<Rt, Fp, Args>::apply(std::integer_sequence<long unsigned int, Is ...>) const [with long unsigned int ...Is = {0, 1, 2}; typename Requires_detail::requires_impl<((void)(sizeof... (Is)), is_callable_v<typename std::decay<_Arg>::type, Deferred_detail::remove_deferred_t<typename std::decay<_BoundArgs>::type>...>)>::template_error_type_failed_to_meet_requirements_on_template_parameters <anonymous> = nullptr; Rt = Variables<brigand::list<Tags::deriv<{anonymous}::ScalarFieldTag, brigand::integral_constant<long unsigned int, 2>, Frame::Inertial, std::nullptr_t> > >; Fp = Variables<brigand::list<Tags::deriv<{anonymous}::ScalarFieldTag, brigand::integral_constant<long unsigned int, 2>, Frame::Inertial, std::nullptr_t> > > (*)(const Variables<brigand::list<{anonymous}::ScalarFieldTag> >&, const Mesh<2>&, const Tensor<DataVector, brigand::list<brigand::integral_constant<int, 2>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<2, (UpLo)0, Frame::Logical, (IndexType)0>, Tensor_detail::TensorIndexType<2, (UpLo)1, Frame::Inertial, (IndexType)0> > >&); Args = {Deferred<Variables<brigand::list<{anonymous}::ScalarFieldTag> >, std::integral_constant<bool, false> >, Deferred<Mesh<2>, std::integral_constant<bool, false> >, Deferred<Tensor<DataVector, brigand::list<brigand::integral_constant<int, 2>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<2, (UpLo)0, Frame::Logical, (IndexType)0>, Tensor_detail::TensorIndexType<2, (UpLo)1, Frame::Inertial, (IndexType)0> > >, std::integral_constant<bool, false> >}]’
/home/francois/code/spectre-work/spectre/src/DataStructures/DataBox/Deferred.hpp:123:7:   required from ‘const Rt& Deferred_detail::deferred_assoc_state<Rt, Fp, Args>::get() const [with Rt = Variables<brigand::list<Tags::deriv<{anonymous}::ScalarFieldTag, brigand::integral_constant<long unsigned int, 2>, Frame::Inertial, std::nullptr_t> > >; Fp = Variables<brigand::list<Tags::deriv<{anonymous}::ScalarFieldTag, brigand::integral_constant<long unsigned int, 2>, Frame::Inertial, std::nullptr_t> > > (*)(const Variables<brigand::list<{anonymous}::ScalarFieldTag> >&, const Mesh<2>&, const Tensor<DataVector, brigand::list<brigand::integral_constant<int, 2>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<2, (UpLo)0, Frame::Logical, (IndexType)0>, Tensor_detail::TensorIndexType<2, (UpLo)1, Frame::Inertial, (IndexType)0> > >&); Args = {Deferred<Variables<brigand::list<{anonymous}::ScalarFieldTag> >, std::integral_constant<bool, false> >, Deferred<Mesh<2>, std::integral_constant<bool, false> >, Deferred<Tensor<DataVector, brigand::list<brigand::integral_constant<int, 2>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<2, (UpLo)0, Frame::Logical, (IndexType)0>, Tensor_detail::TensorIndexType<2, (UpLo)1, Frame::Inertial, (IndexType)0> > >, std::integral_constant<bool, false> >}]’
/home/francois/code/spectre-work/spectre/src/DataStructures/DataBox/Deferred.hpp:121:13:   required from here
/home/francois/code/spectre-work/spectre/src/DataStructures/Variables.hpp:186:22: internal compiler error: in tsubst_pack_expansion, at cp/pt.c:12200
             Requires<tmpl2::flat_all_v<
                      ^~~~~
[snip]
```

### Environment:

Content of `LibraryVersions.txt`:

```
CMake Version: 3.13.0
SpECTRE Version: 0.0.0
SpECTRE branch: HEAD
SpECTRE hash: 53ae3e859e810f6e59390d7165dc0d065e3ecd54
Hostname: noisette
Host system: Linux-4.19.4-300.fc29.x86_64
Host system version: 4.19.4-300.fc29.x86_64
Host system processor: x86_64
CMAKE_CXX_FLAGS:  -fPIC
CMAKE_CXX_LINK_FLAGS:
CMAKE_CXX_FLAGS_DEBUG: -g
CMAKE_CXX_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_C_FLAGS:  -fPIC
CMAKE_C_LINK_FLAGS:
CMAKE_C_FLAGS_DEBUG: -g
CMAKE_C_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_Fortran_FLAGS:
CMAKE_Fortran_LINK_FLAGS:
CMAKE_Fortran_FLAGS_DEBUG: -g
CMAKE_Fortran_FLAGS_RELEASE: -O3 -DNDEBUG -O3
CMAKE_EXE_LINKER_FLAGS:
CMAKE_BUILD_TYPE: Debug
CMAKE_CXX_COMPILER: /home/francois/spack/opt/spack/linux-fedora29-x86_64/gcc-8.2.1/gcc-8.2.0-bgztg7i4hriklsfwrxnx4r6xgftcuprx/bin/g++
CMAKE_CXX_COMPILER_VERSION: 8.2.0
CMAKE_C_COMPILER: /home/francois/spack/opt/spack/linux-fedora29-x86_64/gcc-8.2.1/gcc-8.2.0-bgztg7i4hriklsfwrxnx4r6xgftcuprx/bin/gcc
CMAKE_C_COMPILER_VERSION: 8.2.0
CMAKE_Fortran_COMPILER: /home/francois/spack/opt/spack/linux-fedora29-x86_64/gcc-8.2.1/gcc-8.2.0-bgztg7i4hriklsfwrxnx4r6xgftcuprx/bin/gfortran
CMAKE_Fortran_COMPILER_VERSION: 8.2.0
Python Version: 2.7.15
Charm Version:  6.8.0
Charm SMP:  YES
BLAS_LIBRARIES:  /home/francois/spack/opt/spack/linux-fedora29-x86_64/gcc-8.2.1/openblas-0.3.3-ks56jgo57jivwn4nrq5jhwpswuwqeevo/lib/libopenblas.so
Blaze Version:  3.2
Boost Version:  1.68.0
Catch Version:  2.2.1
GSL Version:  2.5
HDF5 Version:  1.10.4
jemalloc Version:  ..
LIBXSMM Version:  ..
LAPACK_LIBRARIES:  /home/francois/spack/opt/spack/linux-fedora29-x86_64/gcc-8.2.1/openblas-0.3.3-ks56jgo57jivwn4nrq5jhwpswuwqeevo/lib/libopenblas.so;/home/francois/spack/opt/spack/linux-fedora29-x86_64/gcc-8.2.1/openblas-0.3.3-ks56jgo57jivwn4nrq5jhwpswuwqeevo/lib/libopenblas.so
NumPy Version:  1.15.2
SciPy Version:  1.1.0
```

**Comment by hrueter** [comment]
This could be this bug:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85214
